kernel/aster-nix/src/process/posix_thread/futex.rs

```diff
@@ -32,22 +34,23 @@ pub fn futex_wait_bitset(
         "futex_wait_bitset addr: {:#x}, val: {}, timeout: {:?}, bitset: {:#x}",
         futex_addr, futex_val, timeout, bitset
     );
-    let futex_key = FutexKey::new(futex_addr);
-    let (_, futex_bucket_ref) = get_futex_bucket(futex_key);
+    let futex_key = FutexKey::new(futex_addr, bitset);
+    if futex_key.load_val() != futex_val {
+        return_errno_with_message!(Errno::EAGAIN, "futex value does not match");
+    }
+    let (waiter, futex_item) = FutexItem::build(futex_key);
 
     // lock futex bucket ref here to avoid data race
+    let (_, futex_bucket_ref) = get_futex_bucket(futex_key);
     let mut futex_bucket = futex_bucket_ref.lock();
 
-    if futex_key.load_val() != futex_val {
```

Comment:
Notice that in the original implementation, this comparison is with the critical region protected by the lock on the futex bucket. But in the new implementation, the comparison is moved out of the critical region, which is wrong. I don't have the time to show you an execution sequence that triggers such a race condition. I think you can figure it out yourself.

Hint: consider the following futex-based mutex.

```
mutex_init:
    is_locked = 0

mutex_lock:
    while is_locked.swap(1) {
        futex_wait(&is_locked, 1)
    }

mutex_unlock:
    is_locked.set(0)
    futex_wake(&is_locked)
```